nixos/modules/services/hardware/amdgpu.nix

```diff
@@ -0,0 +1,42 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.hardware.amdgpu;
+in {
+  options.hardware.amdgpu = {
+    legacySupport.enable = lib.mkEnableOption ''support for old AMD graphics cards,
+      this uses `amdgpu` instead of `radeonsi` driver for Southern Islands (Radeon HD 7000)
+      series and Sea Islands (Radeon HD 7000) series cards.
+      Note: this does NOT add support for analog video output graphics cards.
+    '';
+    initrd.enable = lib.mkEnableOption ''loading `amdgpu` kernelModule in stage 1.
+      Enable this to fix lower resolution in boot screen during initramfs phase.
+    '';
+    rocm.enable = lib.mkEnableOption ''ROCm/OpenCL support'';
+    # cfg.amdvlk option is defined in ./amdvlk.nix module
+  };
+
+  config = {
+    boot.kernelParams = lib.optionals cfg.legacySupport.enable [
+      "amdgpu.si_support=1"
+      "amdgpu.cik_support=1"
+      "radeon.si_support=0"
+      "radeon.cik_support=0"
+    ];
+
+    boot.initrd.kernelModules = lib.optionals cfg.initrd.enable [ "amdgpu" ];
+
+    hardware.opengl = lib.mkIf cfg.rocm.enable {
+      enable = lib.mkDefault true;
```

Comment:
Why is this being done here? I don't think ROCm requires any of this?

Reply:
Isn't the general idea to use ROCm packages in `hardware.opengl.extraPackages`? That exposes it in `/run/opengl-driver`

Reply:
yes. when `hardware.opengl.enable = false`, then `hardware.opengl.extraPackages` becomes a no-op.

Reply:
I'd note that this is the way `nixos-hardware` [does it](https://github.com/NixOS/nixos-hardware/blob/cde8f7e11f036160b0fd6a9e07dc4c8e4061cf06/common/gpu/amd/default.nix#L28) and [the wiki](https://nixos.wiki/wiki/AMD_GPU).

Reply:
No it doesn't? It doesn't enable anything by default when just ROCm is enabled.

Reply:
`hardware.amdgpu.opencl.enable` seems to be more fitting to what it does.

Reply:
Agreed, though I chose `.rocm` so that it doesn't cause conflicts with current nixos-hardware implementation.

Reply:
I think as long as we don't change the actual type of the option, two lib.mkOption actually should merge. So that is hopefully not an issue.